components/utils.py

```diff
@@ -310,26 +336,72 @@ def yum_install(source):
         source_name = subprocess.check_output(
             ['rpm', '-qp', source_path]).strip()
 
-    ctx.logger.info('Checking whether {0} is already installed...'.format(
-        source_path.rstrip('\n\r')))
-    installed = run(['rpm', '-q', source_name], ignore_failures=True)
-    if installed.returncode == 0:
-        ctx.logger.info('Package {0} is already installed.'.format(source))
-        return
+        ctx.logger.info('Checking whether {0} is already installed...'.format(
+            source_path))
+        installed_rpm = _get_installed_rpm_source(source_name)
+        if installed_rpm == source_name:
+            ctx.logger.info('Package {0} is already installed.'.format(source))
+            return
+
+        if is_upgrade() or is_rollback():
+            _remove_existing_rpm_package(source_name)
 
     ctx.logger.info('yum installing {0}...'.format(source_path))
     sudo(['yum', 'install', '-y', source_path])
 
 
+def _remove_existing_rpm_package(source_name):
+    package_name, package_version = _get_rpm_package_details(source_name)
+    existing_src_name = _get_installed_rpm_source(package_name)
+    if existing_src_name:
+        existing_package_name, existing_pkg_version = \
+            _get_rpm_package_details(existing_src_name)
+        ctx.logger.info('Version {0} is currently installed for package {1}. '
+                        'Uninstalling to allow upgrade to version {2}'
+                        .format(existing_pkg_version, existing_src_name,
+                                package_version))
+        sudo(['rpm', '--noscripts', '-e', existing_src_name])
+
+
+def _get_installed_rpm_source(name):
+    '''
+    returns the rpm full source name by the package/source name. None if
+    not found.
+    :param name: package/source name
+    :return: The full source name if installed, else None.
+    '''
+    installed = run(['rpm', '-q', name], ignore_failures=True)
+    if installed.returncode == 0:
+        existing_rpm_pkg_name = installed.aggr_stdout.rstrip('\n\r')
+        ctx.logger.debug('Found installed package named: {0}'
+                         .format(existing_rpm_pkg_name))
+        return existing_rpm_pkg_name
+    return None
+
+
+def _get_rpm_package_details(source_name):
```

Comment:
if a user puts the rpm within their yum repo instead of providing a url to a file, this will not work. a lot of customers want to put the rpms we provide within their yum repos and then just provide a name of the package in the blueprint.

Additionally, we can't assume that the name of the file is indicative of its version and real package name even if the rpm conventions declare so.

rpm/yum provide an interface (horrible one though robest) for getting the real package name and version (https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Deployment_Guide/sec-Displaying_Package_Information.html). This prevents ambiguities and future problems with identification.